File: demos/ruby/basic_proxy.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# Outbound LB - Ruby Demo
#
# Demonstrates how to use Outbound LB proxy with Ruby using net/http.
# Includes examples for HTTP, HTTPS, authentication, error handling, and concurrent requests.
#
# Requirements:
#     Ruby 2.7+
#
# Usage:
#     ruby basic_proxy.rb
#

require 'net/http'
require 'uri'
require 'json'
require 'timeout'

# Configuration from environment variables
PROXY_HOST = ENV.fetch('PROXY_HOST', 'localhost')
PROXY_PORT = ENV.fetch('PROXY_PORT', '3128').to_i
PROXY_USER = ENV.fetch('PROXY_USER', 'user')
PROXY_PASS = ENV.fetch('PROXY_PASS', 'password')

def print_separator(title)
  puts '=' * 60
  puts title
  puts '=' * 60
end

# Example 1: Basic HTTP Request
def example_http_request
  print_separator('Example 1: Basic HTTP Request')

  begin
    uri = URI('http://httpbin.org/ip')

    http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT)
    http.open_timeout = 10
    http.read_timeout = 10

    response = http.get(uri.request_uri)

    puts "Status: #{response.code}"
    puts "Response: #{response.body}"
  rescue StandardError => e
    puts "Error: #{e.message}"
  end

  puts
end

# Example 2: HTTPS Request (CONNECT tunnel)
def example_https_request
  print_separator('Example 2: HTTPS Request (CONNECT tunnel)')

  begin
    uri = URI('https://httpbin.org/ip')

    http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_PEER
    http.open_timeout = 10
    http.read_timeout = 10

    response = http.get(uri.request_uri)

    puts "Status: #{response.code}"
    puts "Response: #{response.body}"
  rescue StandardError => e
    puts "Error: #{e.message}"
  end

  puts
end

# Example 3: Authenticated Proxy
def example_authenticated_proxy
  print_separator('Example 3: Authenticated Proxy')

  begin
    uri = URI('https://httpbin.org/ip')

    http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_PEER
    http.open_timeout = 10
    http.read_timeout = 10

    response = http.get(uri.request_uri)

    puts "Status: #{response.code}"
    puts "Response: #{response.body}"
  rescue StandardError => e
    puts "Error: #{e.message}"
  end

  puts
end

# Example 4: Error Handling
def example_error_handling
  print_separator('Example 4: Error Handling')

  # Test connection timeout
  puts 'Testing connection timeout...'
  begin
    uri = URI('http://httpbin.org/delay/5')

    http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT)
    http.open_timeout = 2 # Short timeout
    http.read_timeout = 2

    http.get(uri.request_uri)
  rescue Net::OpenTimeout, Net::ReadTimeout, Timeout::Error
    puts '  Caught timeout error (expected)'
  rescue StandardError => e
    puts "  Error: #{e.class.name}"
  end

  # Test invalid URL
  puts 'Testing invalid URL...'
  begin
    uri = URI('http://invalid.invalid.invalid')

    http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT)
    http.open_timeout = 5
    http.read_timeout = 5

    http.get(uri.request_uri)
  rescue SocketError, Errno::ECONNREFUSED => e
    puts "  Caught connection error (expected): #{e.class.name}"
  rescue StandardError => e
    puts "  Error: #{e.class.name}"
  end

  # Test proxy auth failure
  puts 'Testing proxy auth failure (if auth required on proxy)...'
  begin
    uri = URI('http://httpbin.org/ip')

    http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT, 'wronguser', 'wrongpass')
    http.open_timeout = 5
    http.read_timeout = 5

    response = http.get(uri.request_uri)

    if response.code == '407'
      puts '  Got 407 Proxy Authentication Required (expected)'
    else
      puts "  Status: #{response.code}"
    end
  rescue StandardError => e
    puts "  Error: #{e.message}"
  end

  puts
end

# Example 5: Concurrent Requests (Load Balancing Demo)
def example_concurrent_requests
  print_separator('Example 5: Concurrent Requests (Load Balancing Demo)')

  num_requests = 10
  puts "Making #{num_requests} concurrent requests..."

  threads = []
  results = []
  mutex = Mutex.new

  num_requests.times do |i|
    threads << Thread.new(i) do |request_id|
      begin
        uri = URI('http://httpbin.org/ip')

        http = Net::HTTP.new(uri.host, uri.port, PROXY_HOST, PROXY_PORT)
        http.open_timeout = 10
        http.read_timeout = 10

        response = http.get(uri.request_uri)
        data = JSON.parse(response.body)
        ip = data['origin']

        mutex.synchronize do
          results << { request_id: request_id, ip: ip }
        end
      rescue StandardError => e
        mutex.synchronize do
          results << { request_id: request_id, ip: "error: #{e.message}" }
        end
      end
    end
  end

  threads.each(&:join)

  # Print results
  results.sort_by { |r| r[:request_id] }.each do |r|
    puts "  Request #{r[:request_id]}: #{r[:ip]}"
  end

  # Count IP distribution
  ip_counts = {}
  results.each do |r|
    next if r[:ip].start_with?('error')

    ip_counts[r[:ip]] ||= 0
    ip_counts[r[:ip]] += 1
  end

  puts "\nIP Distribution:"
  ip_counts.sort.each do |ip, count|
    puts "  #{ip}: #{count} requests"
  end

  puts
end

# Example 6: Using environment variables
def example_environment_variables
  print_separator('Example 6: Using Environment Variables')

  puts 'Ruby respects HTTP_PROXY and HTTPS_PROXY environment variables.'
  puts 'Set them before running your script:'
  puts
  puts '  export HTTP_PROXY=http://localhost:3128'
  puts '  export HTTPS_PROXY=http://localhost:3128'
  puts
  puts 'Then use Net::HTTP without explicit proxy configuration:'
  puts
  puts '  uri = URI("https://api.example.com")'
  puts '  response = Net::HTTP.get_response(uri)'
  puts
end

# Main
def main
  puts
  puts 'Outbound LB - Ruby Demo'
  puts "Proxy: http://#{PROXY_HOST}:#{PROXY_PORT}"
  puts

  example_http_request
  example_https_request
  example_authenticated_proxy
  example_error_handling
  example_concurrent_requests
  example_environment_variables

  puts 'All examples completed!'
end

main if __FILE__ == $PROGRAM_NAME

File: demos/php/basic_proxy.php
#!/usr/bin/env php
<?php
/**
 * Outbound LB - PHP Demo
 *
 * Demonstrates how to use Outbound LB proxy with PHP using cURL.
 * Includes examples for HTTP, HTTPS, authentication, error handling, and concurrent requests.
 *
 * Requirements:
 *     PHP 7.4+ with cURL extension
 *
 * Usage:
 *     php basic_proxy.php
 */

// Configuration from environment variables
$PROXY_HOST = getenv('PROXY_HOST') ?: 'localhost';
$PROXY_PORT = getenv('PROXY_PORT') ?: '3128';
$PROXY_USER = getenv('PROXY_USER') ?: 'user';
$PROXY_PASS = getenv('PROXY_PASS') ?: 'password';

$PROXY_URL = "http://{$PROXY_HOST}:{$PROXY_PORT}";

function printSeparator(string $title): void {
    echo str_repeat('=', 60) . "\n";
    echo $title . "\n";
    echo str_repeat('=', 60) . "\n";
}

/**
 * Example 1: Basic HTTP Request
 */
function exampleHttpRequest(): void {
    global $PROXY_HOST, $PROXY_PORT;

    printSeparator('Example 1: Basic HTTP Request');

    $ch = curl_init();

    curl_setopt_array($ch, [
        CURLOPT_URL => 'http://httpbin.org/ip',
        CURLOPT_PROXY => $PROXY_HOST,
        CURLOPT_PROXYPORT => $PROXY_PORT,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 10,
        CURLOPT_CONNECTTIMEOUT => 10,
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);

    curl_close($ch);

    if ($error) {
        echo "Error: {$error}\n";
    } else {
        echo "Status: {$httpCode}\n";
        echo "Response: {$response}\n";
    }

    echo "\n";
}

/**
 * Example 2: HTTPS Request (CONNECT tunnel)
 */
function exampleHttpsRequest(): void {
    global $PROXY_HOST, $PROXY_PORT;

    printSeparator('Example 2: HTTPS Request (CONNECT tunnel)');

    $ch = curl_init();

    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://httpbin.org/ip',
        CURLOPT_PROXY => $PROXY_HOST,
        CURLOPT_PROXYPORT => $PROXY_PORT,
        CURLOPT_HTTPPROXYTUNNEL => true, // Use CONNECT for HTTPS
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 10,
        CURLOPT_CONNECTTIMEOUT => 10,
        CURLOPT_SSL_VERIFYPEER => true,
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);

    curl_close($ch);

    if ($error) {
        echo "Error: {$error}\n";
    } else {
        echo "Status: {$httpCode}\n";
        echo "Response: {$response}\n";
    }

    echo "\n";
}

/**
 * Example 3: Authenticated Proxy
 */
function exampleAuthenticatedProxy(): void {
    global $PROXY_HOST, $PROXY_PORT, $PROXY_USER, $PROXY_PASS;

    printSeparator('Example 3: Authenticated Proxy');

    $ch = curl_init();

    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://httpbin.org/ip',
        CURLOPT_PROXY => $PROXY_HOST,
        CURLOPT_PROXYPORT => $PROXY_PORT,
        CURLOPT_PROXYUSERPWD => "{$PROXY_USER}:{$PROXY_PASS}",
        CURLOPT_HTTPPROXYTUNNEL => true,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 10,
        CURLOPT_CONNECTTIMEOUT => 10,
        CURLOPT_SSL_VERIFYPEER => true,
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);

    curl_close($ch);

    if ($error) {
        echo "Error: {$error}\n";
    } else {
        echo "Status: {$httpCode}\n";
        echo "Response: {$response}\n";
    }

    echo "\n";
}

/**
 * Example 4: Error Handling
 */
function exampleErrorHandling(): void {
    global $PROXY_HOST, $PROXY_PORT;

    printSeparator('Example 4: Error Handling');

    // Test connection timeout
    echo "Testing connection timeout...\n";

    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'http://httpbin.org/delay/5',
        CURLOPT_PROXY => $PROXY_HOST,
        CURLOPT_PROXYPORT => $PROXY_PORT,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 2, // Short timeout
    ]);

    $response = curl_exec($ch);
    $error = curl_error($ch);
    $errno = curl_errno($ch);
    curl_close($ch);

    if ($errno === CURLE_OPERATION_TIMEOUTED) {
        echo "  Caught timeout error (expected)\n";
    } else if ($error) {
        echo "  Error: {$error}\n";
    }

    // Test invalid URL
    echo "Testing invalid URL...\n";

    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'http://invalid.invalid.invalid',
        CURLOPT_PROXY => $PROXY_HOST,
        CURLOPT_PROXYPORT => $PROXY_PORT,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 5,
    ]);

    $response = curl_exec($ch);
    $error = curl_error($ch);
    $errno = curl_errno($ch);
    curl_close($ch);

    if ($error) {
        echo "  Caught connection error (expected): CURLE_{$errno}\n";
    }

    // Test proxy auth failure
    echo "Testing proxy auth failure (if auth required on proxy)...\n";

    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'http://httpbin.org/ip',
        CURLOPT_PROXY => $PROXY_HOST,
        CURLOPT_PROXYPORT => $PROXY_PORT,
        CURLOPT_PROXYUSERPWD => 'wronguser:wrongpass',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 5,
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($httpCode === 407) {
        echo "  Got 407 Proxy Authentication Required (expected)\n";
    } else if ($error) {
        echo "  Error: {$error}\n";
    } else {
        echo "  Status: {$httpCode}\n";
    }

    echo "\n";
}

/**
 * Example 5: Concurrent Requests (Load Balancing Demo)
 */
function exampleConcurrentRequests(): void {
    global $PROXY_HOST, $PROXY_PORT;

    printSeparator('Example 5: Concurrent Requests (Load Balancing Demo)');

    $numRequests = 10;
    echo "Making {$numRequests} concurrent requests...\n";

    $multiHandle = curl_multi_init();
    $handles = [];

    // Create handles
    for ($i = 0; $i < $numRequests; $i++) {
        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL => 'http://httpbin.org/ip',
            CURLOPT_PROXY => $PROXY_HOST,
            CURLOPT_PROXYPORT => $PROXY_PORT,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 10,
        ]);

        curl_multi_add_handle($multiHandle, $ch);
        $handles[$i] = $ch;
    }

    // Execute all requests
    $running = null;
    do {
        curl_multi_exec($multiHandle, $running);
        curl_multi_select($multiHandle);
    } while ($running > 0);

    // Collect results
    $ipCounts = [];

    foreach ($handles as $i => $ch) {
        $response = curl_multi_getcontent($ch);
        $error = curl_error($ch);

        if ($error) {
            echo "  Request {$i}: error: {$error}\n";
        } else {
            $data = json_decode($response, true);
            $ip = $data['origin'] ?? 'unknown';
            echo "  Request {$i}: {$ip}\n";

            if (!isset($ipCounts[$ip])) {
                $ipCounts[$ip] = 0;
            }
            $ipCounts[$ip]++;
        }

        curl_multi_remove_handle($multiHandle, $ch);
        curl_close($ch);
    }

    curl_multi_close($multiHandle);

    echo "\nIP Distribution:\n";
    ksort($ipCounts);
    foreach ($ipCounts as $ip => $count) {
        echo "  {$ip}: {$count} requests\n";
    }

    echo "\n";
}

/**
 * Example 6: Using file_get_contents with stream context
 */
function exampleStreamContext(): void {
    global $PROXY_HOST, $PROXY_PORT;

    printSeparator('Example 6: Using Stream Context');

    $context = stream_context_create([
        'http' => [
            'proxy' => "tcp://{$PROXY_HOST}:{$PROXY_PORT}",
            'request_fulluri' => true,
            'timeout' => 10,
        ],
    ]);

    $response = @file_get_contents('http://httpbin.org/ip', false, $context);

    if ($response === false) {
        echo "Error: Failed to fetch URL\n";
    } else {
        echo "Response: {$response}\n";
    }

    echo "\n";
}

// Main
function main(): void {
    global $PROXY_URL;

    echo "\n";
    echo "Outbound LB - PHP Demo\n";
    echo "Proxy: {$PROXY_URL}\n";
    echo "\n";

    // Check if cURL is available
    if (!function_exists('curl_init')) {
        echo "Error: cURL extension is required\n";
        exit(1);
    }

    exampleHttpRequest();
    exampleHttpsRequest();
    exampleAuthenticatedProxy();
    exampleErrorHandling();
    exampleConcurrentRequests();
    exampleStreamContext();

    echo "All examples completed!\n";
}

main();
